builtins: implement swathe of I/O functions for Geometry/Geography

otan opened the pull request:
The main weirdness here is that ST_Geo*fromText can take EWKT and an
SRID and the SRID always overwrites the EWKT value. As such, minor
changes were made to the parser in the geo package.

Release note (sql change): We have added new functionality for the
following functions:
* ST_GeomFromText, ST_GeometryFromText
* ST_GeogFromText, ST_GeographyFromText
* ST_GeomFromWKB
* ST_GeomFromEWKB
* ST_GeomFromEWKT
* ST_GeomFromGeoJSON
* ST_GeogFromWKB
* ST_GeogFromEWKB
* ST_GeogFromEWKT
* ST_GeogFromGeoJSON
* ST_AsText
* ST_AsBinary
* ST_AsEWKB
* ST_AsEWKT
* ST_AsKML
* ST_AsGeoJSON

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48015)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48015#-:-M5xvyYjBYErb48UwOWp:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/parse.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48015#-M5xjVLz3jwZngohJ85q:-M5xjVLz3jwZngohJ85r:b-eg523x) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d49173fa24830b694c2619853b5bc7875617a6/pkg/geo/parse.go#L73)):*
> ```Go
> 	}
> 	// TODO(otan): check SRID is valid against spatial_ref_sys.
> 	if overwrite || (defaultSRID != 0 && t.SRID() == 0) {
> ```

it is odd to define a type using a bool and refer to it as a bool in (read) expressions.
I would slightly prefer
`if overwriteSetting == DefaultSRIDShouldOverwrite ...`

Similar comment for `ParseEWKT()`
___
*[pkg/geo/parse.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48015#-M5xkA192WytoyWTn50c:-M5xkA192WytoyWTn50d:b5g3802) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d49173fa24830b694c2619853b5bc7875617a6/pkg/geo/parse.go#L157)):*
> ```
> Only override the SRID if the SRID is not zero AND we should not overwrite it.
> ```

The use of both override and overwrite in the same sentence is confusing. How about

Only use the parsed SRID if the parsed SRID is not zero and it was not to be overwritten by the `defaultSRID` parameter.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: implement swathe of I/O functions for Geometry/Geography

The main weirdness here is that ST_Geo*fromText can take EWKT and an
SRID and the SRID always overwrites the EWKT value. As such, minor
changes were made to the parser in the geo package.

Release note (sql change): We have added new functionality for the
following functions:
* ST_GeomFromText, ST_GeometryFromText
* ST_GeogFromText, ST_GeographyFromText
* ST_GeomFromWKB
* ST_GeomFromEWKB
* ST_GeomFromEWKT
* ST_GeomFromGeoJSON
* ST_GeogFromWKB
* ST_GeogFromEWKB
* ST_GeogFromEWKT
* ST_GeogFromGeoJSON
* ST_AsText
* ST_AsBinary
* ST_AsEWKB
* ST_AsEWKT
* ST_AsKML
* ST_AsGeoJSON

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48015)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/parse.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48015#-M5xjVLz3jwZngohJ85q:-M5yI1sD5pR4X7tpGIij:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d49173fa24830b694c2619853b5bc7875617a6/pkg/geo/parse.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

it is odd to define a type using a bool and refer to it as a bool in (read) expressions.
I would slightly prefer
`if overwriteSetting == DefaultSRIDShouldOverwrite ...`

Similar comment for `ParseEWKT()`
</blockquote></details>

Done.
___
*[pkg/geo/parse.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48015#-M5xkA192WytoyWTn50c:-M5yI2TqAefdGJh3zWc-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52d49173fa24830b694c2619853b5bc7875617a6/pkg/geo/parse.go#L157)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> ```
> Only override the SRID if the SRID is not zero AND we should not overwrite it.
> ```

The use of both override and overwrite in the same sentence is confusing. How about

Only use the parsed SRID if the parsed SRID is not zero and it was not to be overwritten by the `defaultSRID` parameter.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola

TFTR!
